python/tvm/relay/analysis.py

```diff
@@ -408,3 +411,116 @@ def structural_hash(value):
         msg = ("found value of type {0} expected" +
                "relay.Expr or relay.Type").format(type(value))
         raise TypeError(msg)
+
+def _export_as_relayviz(expr):
+    """Export a Relay function as a nested dictionary, following the RelayViz spec
+    (https://discuss.tvm.ai/t/rfc-visualizing-relay-program-as-graph/4825/10). The dictionary will
+    contain all information useful for visualizing the Relay program and is meant to be consumed
+    by other visualizers.
+
+    Parameters
+    ----------
+    expr : tvm.relay.Expr
+        The input expression.
+
+    Returns
+    -------
+    viz : dict
+        Nested dictionary
+    """
+
+    # node_dict maps a Relay node to an index (node ID)
+    def _traverse_expr(node, node_dict):
+        if node in node_dict:
+            return
+        node_dict[node] = len(node_dict)
+
+    node_dict = {}
+    post_order_visit(expr, lambda x: _traverse_expr(x, node_dict))
+
+    relayviz_nodes = []
+
+    # Sort by node ID
+    for node, node_idx in sorted(node_dict.items(), key=lambda x: x[1]):
+        if isinstance(node, Function):
+            relayviz_nodes.append({
+                'node_kind': 'Function',
+                'body': node_dict[node.body],
+                'params': [node_dict[x] for x in node.params],
+                'ret_type': {
+                    'dtype': node.ret_type.dtype,
+                    'shape': [int(x) for x in node.ret_type.shape]
+                }
+            })
+        elif isinstance(node, Var):
+            relayviz_nodes.append({
+                'node_kind': 'Var',
+                'name': node.name_hint,
+                'dtype': node.type_annotation.dtype,
+                'shape': [int(x) for x in node.type_annotation.shape]
+            })
+        elif isinstance(node, Call):
+            relayviz_nodes.append({
+                'node_kind': 'Call',
+                'op': node_dict[node.op],
+                'args': [node_dict[arg] for arg in node.args]
+            })
+        elif isinstance(node, Op):
+            relayviz_nodes.append({
+                'node_kind': 'Op',
+                'name': node.name,
+                'attrs': {}
+            })
+        elif isinstance(node, TupleGetItem):
+            relayviz_nodes.append({
+                'node_kind': 'TupleGetItem',
+                'tuple_value': node_dict[node.tuple_value],
+                'index': node.index
```

Comment:
Is this the same as an attribute on an op? Why does this node type need to be special? (actually, why does any node type need to be special?)

Reply:
@broune Relay is not a dataflow graph but a programming language. So the current prototype makes differentiation between different kinds of language constructs.

Reply:
This is about providing sufficient information to the visualizer so that it can create a good visualization while being as simple as possible. In this case, I wonder whether the visualizer has a good use of a special case for the index field or TupleGetItem versus other ops. If not, seems simpler to not have the special case for this, and same question on a case-by-case basis for other cases.

Reply:
@broune I see your point. In this case, you're okay with grouping Relay nodes into, say, 3-4 possible categories?

Reply:
Sorry for the late reply. I'm OK with whatever you think makes sense on this; you're more involved in the details of this than I am, and what I most want (and am very OK with!) is to have the ability to see Relay graphs easily visualized. I just wanted to raise this as a thing to consider.